bundles/framework/announcements/view/AnnouncementsHandler.js

```diff
@@ -1,91 +1,72 @@
-import React from 'react';
-import { StateHandler, controllerMixin, Messaging } from 'oskari-ui/util';
-import { Message } from 'oskari-ui';
-import { LOCAL_STORAGE_KEY, LOCAL_STORAGE_SEPARATOR } from './Constants';
+import { StateHandler, controllerMixin } from 'oskari-ui/util';
+import { isUpcoming, isOutdated, isActive } from '../service/util';
 
 // Handler for announcements. Handles state and service calls.
-
-const getMessage = (key, args) => <Message messageKey={key} messageArgs={args} bundleKey='announcements' />;
-
 class ViewHandler extends StateHandler {
     constructor (service) {
         super();
         this.service = service;
-        this.state = {
-            dontShowAgain: [],
-            announcements: [],
-            showAsPopup: []
-        };
-        this.initState();
+        this.service.on('tool', () => this.updateState({ tools: this.service.getTools() }));
+        this.service.on('fetch', () => this.onFetch());
+        this.service.fetchAnnouncements();
     }
 
-    initState () {
-        this.service.fetchAnnouncements(function (err, announcements) {
-            if (err) {
-                Messaging.error(getMessage('messages.getFailed'));
-            } else {
-                const dontShowAgain = this.getDontShowAgainIds(announcements);
-                const showAsPopup = announcements.filter(ann => ann.options.showAsPopup && !dontShowAgain.includes(ann.id));
-                this.updateState({
-                    announcements,
-                    dontShowAgain,
-                    showAsPopup
-                });
-            }
-        }.bind(this));
-    }
+    onFetch () {
+        const allAnnouncements = this.service.getAnnouncements();
+        const dontShowAgain = this.service.getIdsFromLocalStorage();
+
+        // Admin gets all announcements
+        const announcements = allAnnouncements.filter(a => isActive(a));
 
-    getDontShowAgainIds (announcements = this.state.announcements) {
-        const storage = this.getIdsFromLocalStorage();
-        const ids = announcements.map(a => a.id);
-        return ids.filter(id => storage.includes(id));
+        const newState = {
+            outdated: allAnnouncements.filter(a => isOutdated(a)),
+            upcoming: allAnnouncements.filter(a => isUpcoming(a)),
+            announcements,
+            dontShowAgain
+        };
+        // Filter active announcements to show in banner or popup, mark banner/popup as shown if list is empty
+        if (!this.state.popupShown) {
+            const ann = announcements.filter(ann => ann.options.showAsPopup && !dontShowAgain.includes(ann.id));
+            newState.showAsPopup = ann;
+            newState.popupShown = !ann.length;
+        }
+        if (!this.state.bannerShown) {
+            const ann = announcements.filter(ann => !ann.options.showAsPopup && !dontShowAgain.includes(ann.id));
+            newState.showAsBanner = ann;
+            newState.bannerShown = !ann.length;
+        }
```

Comment:
I would not include the boolean that can be easily checked from the array lenght, but I'm not sure if there is an alternative use for it? The fetch is only done once so the if will always be true?

I would rename `showAsPopup` as `popupAnnouncements` and `showAsBanner` `bannerAnnouncements`
```suggestion
        if (!this.state.popupShown) {
            const ann = announcements.filter(ann => ann.options.showAsPopup && !dontShowAgain.includes(ann.id));
            newState.popupAnnouncements= ann;
            newState.popupShown = !ann.length;
        }
        if (!this.state.bannerShown) {
            const ann = announcements.filter(ann => !ann.options.showAsPopup && !dontShowAgain.includes(ann.id));
            newState.bannerAnnouncements= ann;
            newState.bannerShown = !ann.length;
        }
```

Reply:
For admin users fetch is used after add and update.

Reply:
Removed boolean. If state doesn't have array populates array on fetch (one time only) and empty array means that popup is shown and closed.